### PR TITLE
chore: refactor resource manager domain

### PR DESCRIPTION
**Pull request by lcpojr**
After a hundred attempts trying to fix CI and discovering that ubuntu 20 LTS is breaking the crypto library i finally did it.
This is a major refactor on resource manager and some api calls in order to be more clean.

### CODE REVIEW

File: apps/authorizer/test/support/data_case.ex [hunk 1]
```diff
@@ -13,4 +13,18 @@ defmodule Authorizer.DataCase do
...
+      assert {:error, changeset} = Accounts.create_user(%{password: "short"})
+      assert "password is too short" in errors_on(changeset).password
+      assert %{password: ["password is too short"]} = errors_on(changeset)
+  """
+  def errors_on(changeset) do
+    Ecto.Changeset.traverse_errors(changeset, fn {message, opts} ->
+      Enum.reduce(opts, message, fn {key, value}, acc ->
+        String.replace(acc, "%{#{key}}", to_string(value))
+      end)
+    end)
```
Comment (yashin5) [inline apps/authorizer/test/support/data_case.ex:28]: dont you think be more simple create a other project in umbrella that centralize helpers like that? @lcpojr 
I don't really like creating repeated funcs that way

File: apps/resource_manager/lib/credentials/schemas/password.ex [hunk 2]
```diff
@@ -12,48 +12,100 @@ defmodule ResourceManager.Credentials.Schemas.Password do
...
+  @default_salt 16
+
+  # Changeset validation arguments
+  @acceptable_algorithms ~w(argon2 bcrypt pbkdf2)
 
   @required_fields [:password_hash]
-  @foreign_key_fields [:user_id]
-  @optional_fields [:algorithm]
+  @optional_fields [:user_id, :value, :algorithm, :salt]
   schema "passwords" do
```
Comment (yashin5) [inline apps/resource_manager/lib/credentials/schemas/password.ex:42]: if you set :pasword, :string, virtual: true here, you dont need to validate if password is binary in the below funcitons, right?

File: apps/resource_manager/lib/credentials/schemas/public_key.ex [hunk 3]
```diff
@@ -38,28 +39,25 @@ defmodule ResourceManager.Credentials.Schemas.PublicKey do
...
-    %__MODULE__{}
-    |> cast(params, @required_fields ++ @foreign_key_fields ++ @optional_fields)
-    |> validate_required(@required_fields ++ @foreign_key_fields)
-    |> validate_inclusion(:type, @possible_types)
-    |> validate_inclusion(:format, @possible_formats)
-    |> unique_constraint(:user_id)
-  end
+  @doc "Generates an `%Ecto.Changeset{}` to be used in insert operations"
+  @spec changeset(params :: map()) :: Ecto.Changeset.t()
+  def changeset(params) when is_map(params), do: changeset(%__MODULE__{}, params)
```
Comment (yashin5) [inline apps/resource_manager/lib/credentials/schemas/public_key.ex:44]:
```suggestion
  def changeset(%{} = params), do: changeset(%__MODULE__{}, params)
```

File: apps/resource_manager/lib/credentials/schemas/public_key.ex [hunk 4]
```diff
@@ -38,28 +39,25 @@ defmodule ResourceManager.Credentials.Schemas.PublicKey do
...
 
-  @doc false
-  def changeset_update(%__MODULE__{} = model, params) when is_map(params) do
+  @doc "Generates an `%Ecto.Changeset to be used in update operations."
+  @spec changeset(model :: __MODULE__.t(), params :: map()) :: Ecto.Changeset.t()
+  def changeset(%__MODULE__{} = model, params) when is_map(params) do
```
Comment (yashin5) [inline apps/resource_manager/lib/credentials/schemas/public_key.ex:48]:
```suggestion
  def changeset(%__MODULE__{} = model, %{} = params) do
```

File: apps/resource_manager/lib/identities/commands/create_client_application.ex [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        {:ok, app}
+      else
+        {:error, reason} = result ->
+          Logger.error("Failed to create client application", reason: reason)
+          result
+      end
+    end)
+  end
+
+  def execute(params) when is_map(params) do
```
Comment (yashin5) [inline apps/resource_manager/lib/identities/commands/create_client_application.ex:35]:
```suggestion
  def execute(%{} = params)  do
```

File: apps/resource_manager/lib/identities/commands/create_user.ex [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  @typedoc "All possible responses"
+  @type possible_response :: {:ok, User.t()} | {:error, Ecto.Changeset.t()}
+
+  @doc "Create a new identity with it's credentials"
+  @spec execute(params :: CreateUser.t()) :: possible_response()
+  def execute(%CreateUser{} = input) do
+    Logger.debug("Creating new user identity")
+
+    Repo.execute_transaction(fn ->
```
Comment (yashin5) [inline apps/resource_manager/lib/identities/commands/create_user.ex:22]: Repo.execute_transaction in like Repo.transaction?

File: apps/resource_manager/lib/identities/commands/create_user.ex [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        {:ok, user}
+      else
+        {:error, reason} = result ->
+          Logger.error("Failed to create user", reason: reason)
+          result
+      end
+    end)
+  end
+
+  def execute(params) when is_map(params) do
```
Comment (yashin5) [inline apps/resource_manager/lib/identities/commands/create_user.ex:35]:
```suggestion
  def execute(%{} = params)  do
```

File: apps/resource_manager/lib/identities/schemas/client_application.ex [hunk 8]
```diff
@@ -56,50 +56,47 @@ defmodule ResourceManager.Identities.Schemas.ClientApplication do
...
-  @doc false
-  def changeset_create(params) when is_map(params) do
-    %__MODULE__{}
+  @doc "Generates an `%Ecto.Changeset{}` to be used in insert operations"
+  @spec changeset(params :: map()) :: Ecto.Changeset.t()
+  def changeset(params) when is_map(params), do: changeset(%__MODULE__{}, params)
+
+  @doc "Generates an `%Ecto.Changeset to be used in update operations."
+  @spec changeset(model :: __MODULE__.t(), params :: map()) :: Ecto.Changeset.t()
+  def changeset(%__MODULE__{} = model, params) when is_map(params) do
```
Comment (yashin5) [inline apps/resource_manager/lib/identities/schemas/client_application.ex:65]:
```suggestion
  def changeset(%{} = params) , do: changeset(%__MODULE__{}, params)

  @doc "Generates an `%Ecto.Changeset to be used in update operations."
  @spec changeset(model :: __MODULE__.t(), params :: map()) :: Ecto.Changeset.t()
  def changeset(%__MODULE__{} = model, %{} = params) do
```

File: apps/resource_manager/lib/identities/schemas/user.ex [hunk 9]
```diff
@@ -41,27 +42,25 @@ defmodule ResourceManager.Identities.Schemas.User do
...
-  end
+  @doc "Generates an `%Ecto.Changeset{}` to be used in insert operations"
+  @spec changeset(params :: map()) :: Ecto.Changeset.t()
+  def changeset(params) when is_map(params), do: changeset(%__MODULE__{}, params)
 
-  @doc false
-  def changeset_update(%__MODULE__{} = model, params) when is_map(params) do
+  @doc "Generates an `%Ecto.Changeset to be used in update operations."
+  @spec changeset(model :: __MODULE__.t(), params :: map()) :: Ecto.Changeset.t()
+  def changeset(%__MODULE__{} = model, params) when is_map(params) do
```
Comment (yashin5) [inline apps/resource_manager/lib/identities/schemas/user.ex:51]:
```suggestion
  def changeset(%__MODULE__{} = model, %{} = params)  do
```

File: apps/resource_manager/lib/input.ex [hunk 10]
```diff
@@ -42,18 +42,26 @@ defmodule ResourceManager.Input do
...
-      defp do_cast_to_map([_ | _] = list), do: Enum.map(list, &do_cast_to_map/1)
-
-      defp do_cast_to_map(%_struct{} = struct) do
-        struct
-        |> Map.from_struct()
-        |> Map.new(&do_cast_to_map/1)
+      defp do_cast_to_map({key, %Date{} = value}), do: {key, value}
+      defp do_cast_to_map({key, %DateTime{} = value}), do: {key, value}
+      defp do_cast_to_map({key, %NaiveDateTime{} = value}), do: {key, value}
+      defp do_cast_to_map({key, %Time{} = value}), do: {key, value}
```
Comment (yashin5) [inline apps/resource_manager/lib/input.ex:48]:
```suggestion
      Enum.each([%Date{}, %DateTime{}, %NaiveDateTime{}, %Time{}], fn struct ->
          defp do_cast_to_map({key, unquote(struct) = value}), do: {key, value}
      end)
```

File: apps/resource_manager/lib/input.ex [hunk 11]
```diff
@@ -42,18 +42,26 @@ defmodule ResourceManager.Input do
...
+      defp do_cast_to_map({key, %{__struct__: _} = struct}) do
+        result =
+          struct
+          |> Map.from_struct()
+          |> Map.new(&do_cast_to_map/1)
+
+        {key, result}
       end
 
+      defp do_cast_to_map({key, params}) when is_map(params),
```
Comment (yashin5) [inline apps/resource_manager/lib/input.ex:59]:
```suggestion
      defp do_cast_to_map({key, %{} = params}) ,
```

File: apps/resource_manager/lib/permissions/schema/client_application_scope.ex [hunk 12]
```diff
@@ -30,12 +32,18 @@ defmodule ResourceManager.Permissions.Schemas.ClientApplicationScope do
...
-  @doc false
-  def changeset_create(%__MODULE__{}, params) do
-    %__MODULE__{}
+  @doc "Generates an `%Ecto.Changeset{}` to be used in insert operations"
+  @spec changeset(params :: map()) :: Ecto.Changeset.t()
+  def changeset(params) when is_map(params), do: changeset(%__MODULE__{}, params)
+
+  @doc "Generates an `%Ecto.Changeset to be used in update operations."
+  @spec changeset(model :: __MODULE__.t(), params :: map()) :: Ecto.Changeset.t()
+  def changeset(%__MODULE__{} = model, params) when is_map(params) do
```
Comment (yashin5) [inline apps/resource_manager/lib/permissions/schema/client_application_scope.ex:41]:
```suggestion
  def changeset(%__MODULE__{} = model, %{} = params)  do
```

File: apps/resource_manager/lib/permissions/schema/scope.ex [hunk 13]
```diff
@@ -29,19 +29,16 @@ defmodule ResourceManager.Permissions.Schemas.Scope do
...
-  def changeset_create(params) when is_map(params) do
-    %__MODULE__{}
-    |> cast(params, @required ++ @optional)
-    |> validate_required(@required)
-    |> validate_length(:name, min: 1)
-    |> unique_constraint(:name)
-  end
+  @doc "Generates an `%Ecto.Changeset{}` to be used in insert operations"
+  @spec changeset(params :: map()) :: Ecto.Changeset.t()
+  def changeset(params) when is_map(params), do: changeset(%__MODULE__{}, params)
```
Comment (yashin5) [inline apps/resource_manager/lib/permissions/schema/scope.ex:34]:
```suggestion
  def changeset(%{} = params) , do: changeset(%__MODULE__{}, params)
```

File: apps/resource_manager/lib/permissions/schema/scope.ex [hunk 14]
```diff
@@ -29,19 +29,16 @@ defmodule ResourceManager.Permissions.Schemas.Scope do
...
 
-  @doc false
-  def changeset_update(%__MODULE__{} = model, params) when is_map(params) do
+  @doc "Generates an `%Ecto.Changeset to be used in update operations."
+  @spec changeset(model :: __MODULE__.t(), params :: map()) :: Ecto.Changeset.t()
+  def changeset(%__MODULE__{} = model, params) when is_map(params) do
```
Comment (yashin5) [inline apps/resource_manager/lib/permissions/schema/scope.ex:38]:
```suggestion
  def changeset(%__MODULE__{} = model, %{} = params)  do
```

File: apps/resource_manager/lib/permissions/schema/user_scope.ex [hunk 15]
```diff
@@ -30,13 +32,31 @@ defmodule ResourceManager.Permissions.Schemas.UserScope do
     timestamps()
   end
 
-  @doc false
-  def changeset_create(%__MODULE__{}, params) do
-    %__MODULE__{}
+  @doc "Generates an `%Ecto.Changeset{}` to be used in insert operations"
+  @spec changeset(params :: map()) :: Ecto.Changeset.t()
+  def changeset(params) when is_map(params), do: changeset(%__MODULE__{}, params)
```
Comment (yashin5) [inline apps/resource_manager/lib/permissions/schema/user_scope.ex:37]:
```suggestion
  def changeset(params) , do: changeset(%__MODULE__{}, params)
```

File: apps/resource_manager/lib/permissions/schema/user_scope.ex [hunk 16]
```diff
@@ -30,13 +32,31 @@ defmodule ResourceManager.Permissions.Schemas.UserScope do
...
+
+  @doc "Generates an `%Ecto.Changeset to be used in update operations."
+  @spec changeset(model :: __MODULE__.t(), params :: map()) :: Ecto.Changeset.t()
+  def changeset(%__MODULE__{} = model, params) when is_map(params) do
```
Comment (yashin5) [inline apps/resource_manager/lib/permissions/schema/user_scope.ex:41]:
```suggestion
  def changeset(%__MODULE__{} = model, %{} = params)  do
```

File: apps/authorizer/test/support/data_case.ex (hunk 1 above)
Comment (lcpojr) [reply apps/authorizer/test/support/data_case.ex:28]: Actually that was my fault. We could use here the one from `resource_manager` because these applications depends on it.

File: apps/resource_manager/lib/credentials/schemas/public_key.ex (hunk 3 above)
Comment (lcpojr) [reply apps/resource_manager/lib/credentials/schemas/public_key.ex:44]: I prefer to use guard when it's an map or list and the content itself doesen't matter.
But this is just personal preference hahaha.

File: apps/resource_manager/lib/identities/commands/inputs/create_user.ex [hunk 17]
```diff
@@ -5,40 +5,63 @@ defmodule ResourceManager.Identities.Commands.Inputs.CreateUser do
...
 
   @typedoc "Create user input fields"
   @type t :: %__MODULE__{
           username: String.t(),
           status: String.t(),
-          password_hash: String.t(),
-          password_algorithm: String.t(),
-          scopes: list(String.t()) | nil
+          password: %{
+            value: String.t(),
```
Comment (dcdourado) [inline apps/resource_manager/lib/identities/commands/inputs/create_user.ex:16]:
```suggestion
            hash: String.t(),
```

Comment (dcdourado) [reply apps/resource_manager/lib/identities/commands/inputs/create_user.ex:16]: We may don't want to lose the explicit information that the value stored is hashed, not plain

File: apps/resource_manager/lib/identities/commands/inputs/create_user.ex [hunk 18]
```diff
@@ -5,40 +5,63 @@ defmodule ResourceManager.Identities.Commands.Inputs.CreateUser do
...
   embedded_schema do
     # Identity
     field :username, :string
     field :status, :string, default: "active"
 
-    # Credentials
-    field :password_hash, :string
-    field :password_algorithm, :string, default: "argon2"
+    embeds_one :password, Credential, primary_key: false do
+      field :value, :string
```
Comment (dcdourado) [inline apps/resource_manager/lib/identities/commands/inputs/create_user.ex:33]:
```suggestion
      field :hash, :string
```

File: apps/resource_manager/lib/identities/commands/inputs/create_user.ex [hunk 19]
```diff
@@ -5,40 +5,63 @@ defmodule ResourceManager.Identities.Commands.Inputs.CreateUser do
...
-    |> validate_length(:password_algorithm, min: 1)
-    |> validate_inclusion(:status, User.possible_statuses())
+    |> validate_inclusion(:status, User.acceptable_statuses())
     |> validate_required(@required)
   end
+
+  defp changeset_password(model, params) do
+    model
+    |> cast(params, [:value, :algorithm, :salt])
+    |> validate_length(:value, min: 6)
```
Comment (dcdourado) [inline apps/resource_manager/lib/identities/commands/inputs/create_user.ex:57]: I'm afraid this magic number is going to get kind of hidden... Could we create a getter for this value like `Password.minimum_length()` ?

File: apps/resource_manager/lib/identities/schemas/user.ex [hunk 20]
```diff
@@ -15,17 +15,18 @@ defmodule ResourceManager.Identities.Schemas.User do
 
   @typedoc "User schema fields"
   @type t :: %__MODULE__{
-          id: binary(),
+          id: Ecto.UUID.t(),
           username: String.t(),
           status: String.t(),
-          password: Password.t(),
+          password: Password.t() | Ecto.Association.NotLoaded.t(),
```
Comment (dcdourado) [inline apps/resource_manager/lib/identities/schemas/user.ex:21]: Is this change really needed?

File: apps/resource_manager/lib/resource_manager.ex [hunk 21]
```diff
@@ -4,11 +4,14 @@ defmodule ResourceManager do
...
+  alias ResourceManager.Identities.Commands.{CreateClientApplication, CreateUser, GetIdentity}
   alias ResourceManager.Permissions.Commands.{ConsentScope, RemoveScope}
 
-  @doc "Delegates to #{CreateIdentity}.execute/1"
-  defdelegate create_identity(input), to: CreateIdentity, as: :execute
+  @doc "Delegates to #{CreateUser}.execute/1"
+  defdelegate create_user(input), to: CreateUser, as: :execute
+
+  @doc "Delegates to #{CreateClientApplication}.execute/1"
+  defdelegate create_clien_application(input), to: CreateClientApplication, as: :execute
```
Comment (dcdourado) [inline apps/resource_manager/lib/resource_manager.ex:14]:
```suggestion
  defdelegate create_client_application(input), to: CreateClientApplication, as: :execute
```

File: apps/resource_manager/lib/identities/schemas/user.ex (hunk 20 above)
Comment (lcpojr) [reply apps/resource_manager/lib/identities/schemas/user.ex:21]: Actually this  is just  a more precise way to spec an schema because ecto and repo do not preload the fields by default and the return is exactly this not loaded struct.

File: apps/resource_manager/lib/identities/commands/inputs/create_user.ex (hunk 18 above)
Comment (lcpojr) [reply apps/resource_manager/lib/identities/commands/inputs/create_user.ex:33]: In this case it is the password value and not it's hash.

File: apps/resource_manager/lib/identities/commands/inputs/create_user.ex (hunk 17 above)
Comment (lcpojr) [reply apps/resource_manager/lib/identities/commands/inputs/create_user.ex:16]: In this case it's the play password that we receive and not the hashed  one.
